WIP: Add zoom.us video conferencing support

skunkmb opened the pull request:
This is a WIP commit for #10979. 

**Currently implemented:**

- [x] Adding settings for Zoom credentials.
- [x] Updating backend model for Zoom.
- [x] Endpoint for Zoom URL creation.
- [x] Inserting Zoom URL into message.

**To-do:**

 - [x] Verification of credential inputs (other than just if they are empty).
 - [x] Fix an issue with credentials in settings not updating when re-loaded.
 - [x] Rebasing against `master`.
 - [x] Add tests for Zoom.

**GIF of inserting URL:**

![Zoom GIF](https://user-images.githubusercontent.com/17259768/50531386-a7186000-0abd-11e9-9027-367be339a7fa.gif)

timabbott commented:
I did a rebase + migration rename for the first commit.  I think the backend commit is looking pretty close to mergable; what I'd do to complete it is:
* Add unit tests for both the `realm` API and the; you can use `httpretty` to mock the HTTP response results.  
* I think for validating the provided credentials, we should perhaps just call `get_zoom_video_call_url` and not return the URL; since that confirms end-to-end that we can use that endpoint.  (Could potentially do that passing in a hand-patched copy of the `realm` object).

timabbott commented:
Also, just wanted to say this is really nice initial work :)

timabbott commented:
@shubhamdhama do you have ideas for how to make the settings UI part of this cleaner?

shubhamdhama commented on static/js/settings_org.js:
```diff
@@ -450,7 +473,10 @@ function update_dependent_subsettings(property_name) {
     } else if (property_name === 'realm_video_chat_provider' ||
-               property_name === 'realm_google_hangouts_domain') {
+               property_name === 'realm_google_hangouts_domain' ||
+               property_name === 'realm_zoom_user_id' ||
```
`property_name.startsWith('realm_zoom')` could be better?

timabbott replied:
yeah, I like that.

shubhamdhama commented on static/js/settings_org.js:
```diff
@@ -243,9 +252,23 @@ function set_video_chat_provider_dropdown() {
     $("#id_realm_video_chat_provider").val(chat_provider);
     if (chat_provider === "Google Hangouts") {
         $("#google_hangouts_domain").show();
+        $("#zoom_user_id").hide();
```
I agree with Tim here that we may need `hide_zoom_credentials_form` as well as `show_zoom_credentials_form`. Or just a class can work out for all Zoom related inputs? Something like `$('.zoom_credentials').hide()` / `.show()`?

shubhamdhama replied:
@timabbott which one sounds better, probably latter?

timabbott replied:
They're both reasonable; I guess the latter is probably a good pattern for if/when we add more of these.

skunkmb commented:
@timabbott OK, I've updated the PR to include the backend and frontend tests as well as the correct Zoom verification. Now, if the credentials are not accepted, the following message is shown:

![Zoom credentials error](https://user-images.githubusercontent.com/17259768/50732157-c05a8580-1129-11e9-91ec-f7b3e4465b5b.gif)

I also updated the frontend code to use @shubhamdhama's suggestions. Please let me know if there is anything else that I should update. Thank you!

skunkmb commented on static/js/settings_org.js:
```diff
@@ -831,11 +849,21 @@ exports.build_page = function () {
+            $("#google_hangouts_domain").hide();
+            $("#zoom_user_id").hide();
+            $("#zoom_api_key").hide();
+            $("#zoom_api_secret").hide();
```
@shubhamdhama Good point, I missed that. I've updated it to use the `.zoom_credentials` selector. Thanks!

timabbott commented:
Awesome, thanks for building this @skunkmb!  I made a few changes:
* Removed the `console.log` statements from the node tests
* Changed the error message for invalid credentials to be generic and thus potentially avoid additional translation strings for similar things we do in the future.
* Squashed the commits to just a single one adding the feature, since there's no need for the frontend and backend commits to be separate here.  .
* Updated /features and the changelog to advertise this.

and merged this as 9ddadd39f48e375a74302ab671eaa2665f5873a7.

@skunkmb there's one follow-up that seems worth doing: I think we should make the `zerver/views/video_calls.py` method handle the other supported video call providers as well; this would be useful for clients like mobile and terminal where we might not want to to maintain any client-side code in order to support the video calls feature (as the backend currently works, those clients would need to write custom code for the Jitsi/hangouts providers).  Would you be up for adding this in a quick follow-up?

timabbott commented:
Just wanted to add that this ended up really nice and clean :)

martin-huber commented on zerver/lib/actions.py:
```diff
@@ -5287,3 +5288,15 @@ def do_send_realm_reactivation_email(realm: Realm) -> None:
+
+def get_zoom_video_call_url(realm: Realm) -> str:
+    response = request_zoom_video_call_url(
+        realm.zoom_user_id,
```
Using the realm.zoom_user_id here creates all meetings for the user specified in the realm. In an organization this probably is the organization owner ?
Wouldn't it better to use the user_profile.email here and create the meeting for the "me" user ?
At least in our organization, it is so, that all Zulip emails are the same as the Zoom emails.
I did a patch for this, and it is working nice for us.
But I guess a more general soluation either would need to add a "zoom user id" to the userprofile, or at least respond with a comprehensive error message ?
Just a few thoughts to this.

martin-huber replied:
@marco @timabbott  - what do you think about these thoughts ?

We regularily run into the issue, that a zoom meeting already is in progress if another meeting should start, because of the meetings all created for the same Zoom user.

But if you would configure a Zoom admin user you could create the meetings for any other user, for example with the Zulip user email address or an optional future user-specific zoom email address ?

andersk replied:
@martin-huber This code was rewritten completely in Zulip Server 3.0 last year (#11874), and no longer associate a Zoom user with the realm. If you’re still using 2.2.x or earlier, we recommend you [upgrade](https://zulip.readthedocs.io/en/stable/production/upgrade-or-modify.html).

martin-huber replied:
oh, good, thanks for the hint ! We will give it a test asap !
I think, we did not yet update to 3.x because we also will need to update Ubuntu for that ... But let's see.
Thanks!
